lib/mongo/bulk_write.rb

```diff
@@ -278,12 +278,96 @@ def validate_array_filters!(connection)
 
     def validate_hint!(connection)
       if op_combiner.has_hint?
-        if write_concern && !write_concern.acknowledged?
+        if !can_hint?(connection) && write_concern && !write_concern.acknowledged?
           raise Error::UnsupportedOption.hint_error(unacknowledged_write: true)
         elsif !connection.features.update_delete_option_validation_enabled?
           raise Error::UnsupportedOption.hint_error
         end
       end
     end
+
+    # Loop through the requests and check if each operation is allowed to send
+    # a hint for each operation on the given server version.
+    #
+    # For the following operations, the client can send a hint for servers >= 4.2
+    # and for the rest, the client can only send it for 4.4+:
+    #   - updateOne
+    #   - updateMany
+    #   - replaceOne
+    #
+    # @param [ Connection ] connection The connection object.
+    #
+    # @return [ true | false ] Whether the request is able to send hints for
+    #   the current server version.
+    def can_hint?(connection)
+      gte_4_2 = connection.server.description.server_version_gte?('4.2')
+      gte_4_4 = connection.server.description.server_version_gte?('4.4')
+      op_combiner.requests.all? do |req|
+        op = req.keys.first
+        if req[op].keys.include?(:hint)
+          if [:update_one, :update_many, :replace_one].include?(op)
+            gte_4_2
+          else
+            gte_4_4
+          end
+        else
+          true
+        end
+      end
+    end
+
+    # Perform the request document validation required by driver specifications.
+    # This method validates the first key of each update request document to be
+    # an operator (i.e. start with $) and the first key of each replacement
+    # document to not be an operator (i.e. not start with $). The request document
+    # may be invalid without this method flagging it as such (for example an
+    # update or replacement document containing some keys which are operators
+    # and some which are not), in which case the driver expects the server to
+    # fail the operation with an error.
+    #
+    # Raise an ArgumentError if requests is empty.
+    #
+    # @raise [ Error::InvalidUpdateDocument, Error::InvalidReplacementDocument,
+    #   ArgumentError ]
+    #   if the document is invalid.
+    def validate_requests!
```

Comment:
I think we should add a comment here why our fork differs from upstream instead of silently reverting this change:

![image](https://github.com/braze-inc/mongo-ruby-driver/assets/6033/15b374d2-17e0-4cb1-8b10-d670a934095b)

Reply:
added in [69d2199](https://github.com/braze-inc/mongo-ruby-driver/pull/12/commits/69d2199013fd3c422c40032f8eea5077b906abe5)